server/src/main/java/org/opensearch/common/geo/PointParser.java

```diff
@@ -0,0 +1,227 @@
+        this.supportGeoHash = supportGeoHash;
+        this.invalidFieldErrMsg = supportGeoHash
+            ? String.format(
+                Locale.ROOT,
+                "field must be either %s/%s, %s/%s, or %s",
+                fieldNameX,
+                fieldNameY,
+                FN_GEOJSON_TYPE,
+                FN_GEOJSON_COORDS,
+                FN_GEOHASH
+            )
+            : String.format(
+                Locale.ROOT,
+                "field must be either %s/%s, or %s/%s",
+                fieldNameX,
+                fieldNameY,
+                FN_GEOJSON_TYPE,
+                FN_GEOJSON_COORDS
+            );
+    }
+
+    public Point parseObject(final XContentParser parser, final boolean ignoreZValue, final GeoUtils.EffectivePoint effectivePoint)
+        throws IOException {
+        if (parser.currentToken() != XContentParser.Token.START_OBJECT) {
+            throw new OpenSearchParseException("object is expected");
+        }
+
+        parser.nextToken();
+
+        if (parser.currentToken() == XContentParser.Token.END_OBJECT) {
+            throw new OpenSearchParseException(invalidFieldErrMsg);
+        }
+
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+            throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, parser.currentToken());
+        }
+
+        Point point = null;
+        String field = parser.currentName();
+        if (fieldNameX.equals(field) || fieldNameY.equals(field)) {
+            point = parseBasicFields(parser);
+        } else if (supportGeoHash && FN_GEOHASH.equals(field)) {
+            point = parseGeoHashFields(parser, effectivePoint);
+        } else if (FN_GEOJSON_TYPE.equals(field) || FN_GEOJSON_COORDS.equals(field)) {
+            point = parseGeoJsonFields(parser, ignoreZValue);
+        } else {
+            throw new OpenSearchParseException(invalidFieldErrMsg);
+        }
+
+        if (parser.currentToken() != XContentParser.Token.END_OBJECT) {
+            throw new OpenSearchParseException(invalidFieldErrMsg);
+        }
+
+        return point;
+    }
+
+    private Point parseGeoHashFields(final XContentParser parser, final GeoUtils.EffectivePoint effectivePoint) throws IOException {
+        if (parser.currentToken() != XContentParser.Token.FIELD_NAME) {
+            throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, parser.currentToken());
+        }
+
+        String field = parser.currentName();
+        if (!FN_GEOHASH.equals(field)) {
+            throw new OpenSearchParseException(invalidFieldErrMsg);
+        }
+
+        parser.nextToken();
+
+        if (parser.currentToken() != XContentParser.Token.VALUE_STRING) {
+            throw new OpenSearchParseException("{} must be a string", FN_GEOHASH);
+        }
+
+        String geoHash = parser.text();
+        GeoPoint geoPoint = new GeoPoint();
+        geoPoint = geoPoint.parseGeoHash(geoHash, effectivePoint);
+
+        parser.nextToken();
```

Comment:
nit: may be move this outside the method?

Reply:
The parser should move to next token here to be consistent with other methods.